server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1992,70 +1989,54 @@ protected final Releasable acquireEnsureOpenRef() {
 
     /**
      * When called for the first time, puts the engine into a closing state in which further calls to {@link #acquireEnsureOpenRef()} will
-     * fail with an {@link AlreadyClosedException} and waits for all outstanding ensure-open refs to be released, before returning {@code
-     * true}. If called again, returns {@code false} without waiting.
+     * fail with an {@link AlreadyClosedException} and waits for all outstanding ensure-open refs to be released. If called again, returns
+     * without waiting.
      *
-     * @return a flag indicating whether this was the first call or not.
+     * @param listener the action to perform when all open refs are released
      */
-    private boolean drainForClose() {
+    private void drainForClose(ActionListener<Void> listener) {
         if (isClosing.compareAndSet(false, true) == false) {
             logger.trace("drainForClose(): already closing");
-            return false;
+            return;
```

Comment:
Hmm this seems like it leaks `listener` without completing it. That may be what you intend, but it's pretty trappy. Could we fail it with an `AlreadyClosedException` perhaps?